### PR TITLE
Automatically expand ParameterWithSetpoints when adding results via DataSaver.add_result

### PR DESCRIPTION
**Pull request by jenshnielsen**
To make it easier for end users to use parameter with setpoints and simpler to extend doND functions to support it as well

### CODE REVIEW

File: docs/examples/Parameters/Simple-Example-of-ParameterWithSetpoints.ipynb [hunk 1]
```diff
@@ -501,189 +509,21 @@
    "cell_type": "markdown",
    "metadata": {},
    "source": [
-    "To reduce the amount of typing a simple convenience function is provided that will take a `ParameterWithSetpoints` and get both its values and the setpoints values and return them as tuples for direct insertion into the Dataset."
+    "To reduce the amount of typing if a `ParameterWithSetpoints` is given without its setpoints they will automatically be fethed, get called on them and added to the results."
```
Comment (astafan8) [inline docs/examples/Parameters/Simple-Example-of-ParameterWithSetpoints.ipynb:512]:
```suggestion
    "To reduce the amount of typing, if a `ParameterWithSetpoints` is given without its setpoints, the setpoints will be fetched automatically, ``get`` will be called on them, and the obtained data will be added to the results."
```

File: qcodes/dataset/measurements.py [hunk 2]
```diff
@@ -148,6 +160,28 @@ def add_result(self, *res_tuple: res_type) -> None:
...
+        local_results = {}
+        setpoint_names = tuple(setpoint.full_name for setpoint in parameter.setpoints)
+        expanded = tuple(setpoint_name in parameter_names for setpoint_name in setpoint_names)
+        if all(expanded):
+            local_results.update(
+                self._unpack_partial_result(partial_result))
+        elif any(expanded):
+            raise ValueError(f"Some of the setpoints of {parameter.full_name} "
+                             "were explicitly given but others were not. "
+                             "Either supply all of then or none of them.")
```
Comment (astafan8) [inline qcodes/dataset/measurements.py:176]:
```suggestion
                             "Either supply all of them or none of them.")
```

File: qcodes/tests/dataset/test_measurement_context_manager.py [hunk 3]
```diff
@@ -1066,6 +1137,58 @@ def test_datasaver_parameter_with_setpoints_complex(channel_array_instrument,
...
+    assert_allclose(subdata['dummy_channel_inst_ChanA_'
+                            'dummy_parameter_with_setpoints_complex'],
+                    (np.random.rand(n) + 1j * np.random.rand(n)).reshape(1, chan.dummy_n_points()))
+
+
+@pytest.mark.parametrize("bg_writing", [True, False])
+@settings(max_examples=5, deadline=None)
+@given(n=hst.integers(min_value=5, max_value=500))
+@pytest.mark.usefixtures("experiment")
+def test_datasaver_parameter_with_setpoints_complex_expanded(channel_array_instrument,
```
Comment (astafan8) [inline qcodes/tests/dataset/test_measurement_context_manager.py:1167]:
```suggestion
def test_datasaver_parameter_with_setpoints_complex_explicitly_expanded(channel_array_instrument,
```

File: qcodes/tests/dataset/test_measurement_context_manager.py [hunk 4]
```diff
@@ -993,6 +993,77 @@ def test_datasaver_parameter_with_setpoints(channel_array_instrument,
...
+                            'dummy_parameter_with_setpoints'],
+                    expected_data)
+
+
+@settings(max_examples=5, deadline=None)
+@given(n=hst.integers(min_value=5, max_value=500))
+@pytest.mark.parametrize("bg_writing", [True, False])
+@pytest.mark.parametrize("storage_type", ['numeric', 'array'])
+@pytest.mark.usefixtures("experiment")
+def test_datasaver_parameter_with_setpoints_expanded(channel_array_instrument,
```
Comment (astafan8) [inline qcodes/tests/dataset/test_measurement_context_manager.py:1045]:
```suggestion
def test_datasaver_parameter_with_setpoints_explicitly_expanded(channel_array_instrument,
```

File: qcodes/dataset/measurements.py [hunk 5]
```diff
@@ -148,6 +160,28 @@ def add_result(self, *res_tuple: res_type) -> None:
...
+            parameter_names: Sequence[str], partial_result: res_type
+    ) -> Dict[ParamSpecBase, np.ndarray]:
+        local_results = {}
+        setpoint_names = tuple(setpoint.full_name for setpoint in parameter.setpoints)
+        expanded = tuple(setpoint_name in parameter_names for setpoint_name in setpoint_names)
+        if all(expanded):
+            local_results.update(
+                self._unpack_partial_result(partial_result))
+        elif any(expanded):
+            raise ValueError(f"Some of the setpoints of {parameter.full_name} "
```
Comment (astafan8) [inline qcodes/dataset/measurements.py:174]: coverage?

**Review by astafan8 — APPROVED** [review]
That's fair! either pass all values for all setoints yourself or let datasaver do that for all of them, no in-between situation. Great!

File: qcodes/dataset/measurements.py (hunk 5 above)
Comment (jenshnielsen) [reply qcodes/dataset/measurements.py:174]: Sure lets add a test for this

**Pull request merged**